src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -1701,16 +1707,48 @@ private IEnumerable<PSTypeName> InferTypesFrom(MemberExpressionAst memberExpress
 
         private static void AddTypesFromMethodCacheEntry(
             DotNetAdapter.MethodCacheEntry methodCacheEntry,
+            IList<ITypeName> genericTypeArguments,
             List<PSTypeName> result,
             bool isInvokeMemberExpressionAst)
         {
             if (isInvokeMemberExpressionAst)
             {
+                Type[] resolvedTypeArguments = null;
+                if (genericTypeArguments is not null)
+                {
+                    resolvedTypeArguments = new Type[genericTypeArguments.Count];
+                    for (int i = 0; i < genericTypeArguments.Count; i++)
+                    {
+                        Type resolvedType = genericTypeArguments[i].GetReflectionType();
+                        if (resolvedType is null)
+                        {
+                            return;
+                        }
+
+                        resolvedTypeArguments[i] = resolvedType;
+                    }
+                }
+
                 foreach (var method in methodCacheEntry.methodInformationStructures)
                 {
-                    if (method.method is MethodInfo methodInfo && !methodInfo.ReturnType.ContainsGenericParameters)
+                    if (method.method is MethodInfo methodInfo)
                     {
-                        result.Add(new PSTypeName(methodInfo.ReturnType));
+                        if (!methodInfo.ReturnType.ContainsGenericParameters)
+                        {
+                            result.Add(new PSTypeName(methodInfo.ReturnType));
+                        }
+                        else if (genericTypeArguments is not null)
+                        {
+                            try
+                            {
+                                result.Add(new PSTypeName(methodInfo.MakeGenericMethod(resolvedTypeArguments).ReturnType));
+                            }
+                            catch
+                            {
+                                // If we can't build the generic method then just skip it to retain other completion results.
+                                return;
```

Comment:
I've changed it because it made sense with the other changes I made, but why do we want fewer returns in a method? I understand that it can increase the complexity in huge methods with a ton of different paths but I don't see the point in a small and simple method like this one.

Reply:
Just for less places to track when reading code :)